### PR TITLE
Integration tests cleanup

### PR DESCRIPTION
**Pull request by ttuegel**
@virgil-serbanuta Please let me know if this fixes the problems you had on macOS.

See also: #1340 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

**Comment by virgil-serbanuta** [comment]
And here's the MacOs output:
```
~/runtime-verification/kore$ scripts/integration-k.sh 
++ git rev-parse --show-toplevel
+ export TOP=/Users/virgil/runtime-verification/kore
+ TOP=/Users/virgil/runtime-verification/kore
++ stack path --local-bin
+ export PATH=/Users/virgil/runtime-verification/kore/.build/kore/bin:/Users/virgil/.opam/4.06.1+k/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/Users/virgil/runtime-verification/kore/.build/kore/bin:/Users/virgil/.local/bin:/Users/virgil/bin:/Users/virgil/bin/apache-maven-3.5.4/bin/:/Users/virgil/runtime-verification/klab/bin:/Users/virgil/Library/Python/3.7/bin:/Users/virgil/runtime-verification/k/k-distribution/target/release/k/bin
+ PATH=/Users/virgil/runtime-verification/kore/.build/kore/bin:/Users/virgil/.opam/4.06.1+k/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/Users/virgil/runtime-verification/kore/.build/kore/bin:/Users/virgil/.local/bin:/Users/virgil/bin:/Users/virgil/bin/apache-maven-3.5.4/bin/:/Users/virgil/runtime-verification/klab/bin:/Users/virgil/Library/Python/3.7/bin:/Users/virgil/runtime-verification/k/k-distribution/target/release/k/bin
+ MAKE='make --output-sync --jobs --directory /Users/virgil/runtime-verification/kore'
+ make --output-sync --jobs --directory /Users/virgil/runtime-verification/kore test-k
/Applications/Xcode.app/Contents/Developer/usr/bin/make: unrecognized option `--output-sync'
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
  -e, --environment-overrides
                              Environment variables override makefiles.
  -f FILE, --file=FILE, --makefile=FILE
                              Read FILE as a makefile.
  -h, --help                  Print this message and exit.
  -i, --ignore-errors         Ignore errors from commands.
  -I DIRECTORY, --include-dir=DIRECTORY
                              Search DIRECTORY for included makefiles.
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
  -k, --keep-going            Keep going when some targets can't be made.
  -l [N], --load-average[=N], --max-load[=N]
                              Don't start multiple jobs unless load is below N.
  -L, --check-symlink-times   Use the latest mtime between symlinks and target.
  -n, --just-print, --dry-run, --recon
                              Don't actually run any commands; just print them.
  -o FILE, --old-file=FILE, --assume-old=FILE
                              Consider FILE to be very old and don't remake it.
  -p, --print-data-base       Print make's internal database.
  -q, --question              Run no commands; exit status says if up to date.
  -r, --no-builtin-rules      Disable the built-in implicit rules.
  -R, --no-builtin-variables  Disable the built-in variable settings.
  -s, --silent, --quiet       Don't echo commands.
  -S, --no-keep-going, --stop
                              Turns off -k.
  -t, --touch                 Touch targets instead of remaking them.
  -v, --version               Print the version number of make and exit.
  -w, --print-directory       Print the current directory.
  --no-print-directory        Turn off -w, even if it was turned on implicitly.
  -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE
                              Consider FILE to be infinitely new.
  --warn-undefined-variables  Warn when an undefined variable is referenced.
  -N OPTION, --NeXT-option=OPTION
                              Turn on value of NeXT OPTION.

This program built for i386-apple-darwin11.3.0
Report bugs to <bug-make@gnu.org>
```

Fwiw, here's the version information:
```
~/runtime-verification/kore$ make -v
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

I also have 4.2.1, though that's not the default, and if I use it I get the same `make[1]: *** No rule to make target 'test-k'.  Stop.`

**Comment by ttuegel** [comment]
Ah, I forgot to update the targets in the top-level `Makefile`, sorry! It should work now with Make 4. It seems Make 3 doesn't have the `--output-sync` flag; this synchronizes the output so that parallel processes don't write to the terminal over the top of each other. I think it's a very useful option when there is an error. Is depending on Make 4 alright?

**Comment by ttuegel** [comment]
I forgot to add that `make test-k` will _not_ re-run passing tests until you `make clean`, unless you edit the test or definition. This is by design; I found it quite annoying to be re-running unchanged, passing tests all the time. Of course, CI will always run all the tests.

**Comment by virgil-serbanuta** [comment]
> Ah, I forgot to update the targets in the top-level `Makefile`, sorry! It should work now with Make 4. It seems Make 3 doesn't have the `--output-sync` flag; this synchronizes the output so that parallel processes don't write to the terminal over the top of each other. I think it's a very useful option when there is an error. Is depending on Make 4 alright?

I guess so.

Make 3 is the default, installed together with XCode, and it should probably remain the default one. Make 4 is installed by homebrew in a separate directory and needs a path change, which is somewhat annoying, but it's not such a big issue (i.e. I need to run one command per terminal).

**Comment by virgil-serbanuta** [comment]
Fails on my Linux laptop, but I guess it's a z3 issue, I'll check my Mac later, and my Linux desktop after the holidays:
```
~/runtime-verification/kore$ scripts/integration-k.sh 
[...]
/home/virgil/runtime-verification/kore/.build/k/bin/kprove --haskell-backend-command "/home/virgil/runtime-verification/kore/.build/kore/bin/kore-exec " -d . -m VERIFICATION div-spec.k --output-file div-spec.k.out || true
[2019-12-23 10:57:14.041435398][Warning] : [LogMessage] : Failed to decide predicate:
    \and{SortGeneratedTopCell{}}(
        \not{SortGeneratedTopCell{}}(
            /* Created by <makeAndPredicate> at src/Kore/Internal/Predicate.hs:576:32 */
            \and{SortGeneratedTopCell{}}(
                /* Created by <makeAndPredicate> at src/Kore/Internal/Predicate.hs:576:32 */
                \and{SortGeneratedTopCell{}}(
                    \equals{SortInt{}, SortGeneratedTopCell{}}(
                        Lbl'UndsSlsh'Int'Unds'{}(
                            VarX:SortInt{},
                            VarX:SortInt{}
                        ),
                        /* builtin: */ \dv{SortInt{}}("0")
                    ),
                    \not{SortGeneratedTopCell{}}(
                        /* Created by <makeAndPredicate> at src/Kore/Internal/Predicate.hs:576:32 */
                        \and{SortGeneratedTopCell{}}(
                            \equals{SortInt{}, SortGeneratedTopCell{}}(
                                Lbl'UndsSlsh'Int'Unds'{}(
                                    VarX:SortInt{},
                                    VarX:SortInt{}
                                ),
                                /* builtin: */ \dv{SortInt{}}("0")
                            ),
                            \not{SortGeneratedTopCell{}}(
                                \equals{SortInt{}, SortGeneratedTopCell{}}(
                                    VarX:SortInt{},
                                    /* builtin: */ \dv{SortInt{}}("0")
                                )
                            )
                        )
                    )
                ),
                \not{SortGeneratedTopCell{}}(
                    \equals{SortInt{}, SortGeneratedTopCell{}}(
                        VarX:SortInt{},
                        /* builtin: */ \dv{SortInt{}}("0")
                    )
                )
            )
        ),
    \and{SortGeneratedTopCell{}}(
        \not{SortGeneratedTopCell{}}(
            /* Created by <makeAndPredicate> at src/Kore/Internal/Predicate.hs:576:32 */
            \and{SortGeneratedTopCell{}}(
                \equals{SortInt{}, SortGeneratedTopCell{}}(
                    Lbl'UndsSlsh'Int'Unds'{}(VarX:SortInt{}, VarX:SortInt{}),
                    /* builtin: */ \dv{SortInt{}}("0")
                ),
                \not{SortGeneratedTopCell{}}(
                    \equals{SortInt{}, SortGeneratedTopCell{}}(
                        VarX:SortInt{},
                        /* builtin: */ \dv{SortInt{}}("0")
                    )
                )
            )
        ),
    \and{SortGeneratedTopCell{}}(
        \not{SortGeneratedTopCell{}}(
            \equals{SortInt{}, SortGeneratedTopCell{}}(
                Lbl'UndsSlsh'Int'Unds'{}(VarX:SortInt{}, VarX:SortInt{}),
                /* builtin: */ \dv{SortInt{}}("0")
            )
        ),
    \and{SortGeneratedTopCell{}}(
        \not{SortGeneratedTopCell{}}(
            \equals{SortInt{}, SortGeneratedTopCell{}}(
                VarX:SortInt{},
                Lbl'UndsSlsh'Int'Unds'{}(
                    VarX:SortInt{},
                    Lbl'UndsSlsh'Int'Unds'{}(VarX:SortInt{}, VarX:SortInt{})
                )
            )
        ),
        \not{SortGeneratedTopCell{}}(
            \equals{SortInt{}, SortGeneratedTopCell{}}(
                VarX:SortInt{},
                /* builtin: */ \dv{SortInt{}}("0")
            )
        )
    )))) []
diff -u div-spec.k.out.golden div-spec.k.out || ( mv div-spec.k.out div-spec.k.out.fail && false )
--- div-spec.k.out.golden	2019-12-21 01:01:20.105819050 +0200
+++ div-spec.k.out	2019-12-23 10:57:14.426651663 +0200
@@ -1 +1,51 @@
-#True
+  #Not ( #Not ( #Not ( {
+        X
+      #Equals
+        0
+      } )
+    #And
+      {
+        X /Int X
+      #Equals
+        0
+      } )
+  #And
+    #Not ( {
+      X
+    #Equals
+      0
+    } )
+  #And
+    {
+      X /Int X
+    #Equals
+      0
+    } )
+#And
+  #Not ( #Not ( {
+      X
+    #Equals
+      0
+    } )
+  #And
+    {
+      X /Int X
+    #Equals
+      0
+    } )
+#And
+  #Not ( {
+    X
+  #Equals
+    0
+  } )
+#And
+  #Not ( {
+    X /Int X
+  #Equals
+    0
+  } )
+#And
+  <k>
+    X /Int ( X /Int X )
+  </k>
/home/virgil/runtime-verification/kore/test/working/setvars/../include.mk:77: recipe for target 'div-spec.k.out' failed
make[2]: *** [div-spec.k.out] Error 1
make[2]: Leaving directory '/home/virgil/runtime-verification/kore/test/working/setvars'
Makefile:15: recipe for target 'setvars/Makefile.make' failed
make[1]: *** [setvars/Makefile.make] Error 2
Makefile:54: recipe for target 'test-k' failed
make: *** [test-k] Error 2
```

FWIW, here's the z3 version:
```
$ z3 --version
Z3 version 4.4.1
```

**Comment by virgil-serbanuta** [comment]
My Mac:

```
$ stack clean && stack build --pedantic && make clean && scripts/integration-k.sh 
[...]
make --directory bitrange clean
make[2]: Entering directory '/Users/virgil/runtime-verification/kore/test/working/bitrange'
rm -fr ./bitrange-kompiled ./*.out
make[2]: Leaving directory '/Users/virgil/runtime-verification/kore/test/working/bitrange'
make[1]: *** No rule to make target 'Makefile.make', needed by 'clean'.  Stop.
make[1]: Leaving directory '/Users/virgil/runtime-verification/kore/test/working'
make: *** [Makefile:68: clean] Error 2
```

Retrying without `clean`:
```
$ scripts/integration-k.sh 
[...]
make --directory test/working test
make --directory itetest test
make[2]: *** No rule to make target '/Users/virgil/runtime-verification/kore/.build/k/lib/java/kernel-1.0-SNAPSHOT.jar', needed by '1.itetest.out'.  Stop.
make[1]: *** [Makefile:15: itetest/Makefile.make] Error 2
make: *** [Makefile:54: test-k] Error 2
```

Retrying with the frontend (why isn't it an automatic thing?):
```
$ make k-frontend
[...]
$ scripts/integration-k.sh 
[...]
make[2]: Entering directory '/Users/virgil/runtime-verification/kore/test/working/bitrange'
/Users/virgil/runtime-verification/kore/.build/k/bin/krun --haskell-backend-command "/Users/virgil/runtime-verification/kore/.build/kore/bin/kore-exec " -d . 3.bitrange --output-file 3.bitrange.out
diff -u 3.bitrange.out.golden 3.bitrange.out || ( mv 3.bitrange.out 3.bitrange.out.fail && false )
make[2]: Leaving directory '/Users/virgil/runtime-verification/kore/test/working/bitrange'
make[1]: *** No rule to make target 'Makefile.make', needed by 'test-k'.  Stop.
make: *** [Makefile:54: test-k] Error 2
```

**Comment by virgil-serbanuta** [comment]
BTW, I don't expect an answer to these soon, it's just that I had some small chunks of spare time, and starting these tests seemed to fit.

File: test/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  Run the named test and compare the output to `⟨name⟩.⟨ext⟩.out.golden`.
+  The test passes if the expected and actual output are identical.
+  The test will not run again until the test prerequisites change
+  or `⟨name⟩.⟨ext⟩.out` is deleted.
+  If the test fails, the actual output is moved
+  from `⟨name⟩.⟨ext⟩.out` to `⟨name⟩.⟨ext⟩.out.fail`.
+- `make ⟨name⟩.⟨ext⟩.out.golden`:
+  Run the test to construct `⟨name⟩.⟨ext⟩.out.golden`.
+
+## Specification tests
```
Comment (virgil-serbanuta) [inline test/README.md]: Not sure we should add them now, but we also have rule merge tests (https://github.com/kframework/kore/tree/master/src/main/k/working/imp/merge) and tests for saving the proof state (https://github.com/kframework/kore/tree/master/src/main/k/working/imp/save-proofs)

Comment (ttuegel) [reply test/README.md]: Those tests are still there, but I forgot to document them. Thanks!

**Pull request merged**